File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.py
# 백준 1922. 네트워크 연결
# 시간 제한 2초 / 메모리 제한 256MB

import sys
input = sys.stdin.readline

n = int(input()) # 노드 수
m = int(input()) # 간선 수

# union-find 알고리즘
# 1. parent 배열 초기화
parent = [0] * (n+1)
for i in range(1, n+1):
    parent[i] = i

# 2. find 함수
def find(parent, x):
    if parent[x] != x:
        parent[x] = find(parent, parent[x])
    return parent[x]

# 3. union 함수
def union(parent, a, b):
    a = find(parent, a)
    b = find(parent, b)
    if a > b:
        parent[a] = b
    else:
        parent[b] = a

# 크루스칼 알고리즘
edges  = []

for _ in range(m):
    a, b, c = map(int, input().split())

    edges.append([c, a, b])

# 비용을 기준으로 정렬
edges.sort()

total_cost = 0
for i in range(m):
    c, a, b = edges[i]

    if find(parent, a) != find(parent, b):
        union(parent, a, b)
        total_cost += c

print(total_cost)

File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.java
import java.io.*;
import java.util.*;
public class Main {
    static int[] parent;
    static class Edge implements Comparable<Edge> {
        int u, v, w;
        Edge(int u, int v, int w) {
            this.u = u;
            this.v = v;
            this.w = w;
        }
        @Override
        public int compareTo(Edge o) {
            return this.w - o.w;
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        int M = Integer.parseInt(br.readLine());

        List<Edge> edges = new ArrayList<>();

        for(int i=0; i<M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());

            edges.add(new Edge(a, b, c));
        }

        parent = new int[N+1];
        for(int i=1; i<=N; i++) {
            parent[i] = i;
        }

        Collections.sort(edges);

        long ans = 0; int cnt = 0;
        for(Edge e: edges) {
            if(union(e.u, e.v)) {
                ans += e.w;
                if(++cnt==N-1) break;
            }
        }
        System.out.println(ans);
    }

    static boolean union(int a, int b) {
        int rootA = find(a);
        int rootB = find(b);

        if(rootA == rootB) return false;

        parent[rootB] = rootA;
        return true;
    }

    static int find(int x) {
        if(parent[x] != x) {
            parent[x] = find(parent[x]);
        }

        return parent[x];
    }
}